strawberryfields/engine.py

```diff
@@ -336,8 +360,144 @@ def run(self, program, *, shots=1, compile_options={}, modes=None, state_options
+            name (str): The name of the job to be created (e.g. StateTeleportation).
+            shots (int): The number of shots.
+            blackbird_code: The blackbird code of the job.
+
+        Returns:
+            str: A string containing the job content to be sent to the server.
+        """
+        target = self.backend_name
+        return "\n".join(
+            [
+                "name {name}",
+                "version 1.0",
+                "target {target} (shots={shots})",
+                "",
+                "{blackbird_code}",
+            ]
+        ).format(name=name, target=target, shots=str(shots), blackbird_code=blackbird_code)
+
+    def _queue_job(self, job_content):
+        """
+        Create a Job instance based on job_content, and send the job to the API. Append to list
+        of jobs.
+
+        Args:
+            job_content (str): The Blackbird code to execute
+
+        Returns:
+            (strawberryfields.api_client.Job): A Job instance referencing the queued job.
+        """
+        job = Job(client=self.client)
+        job.manager.create(circuit=job_content)
+        self.jobs.append(job)
+        return job
+
+    def _run_program(self, program, **kwargs):
+        """
+        Given a compiled program, gets the blackbird circuit code and creates (or resumes) a job
+        via the API. If the job is completed, returns the job result.
+
+        A queued job can be interrupted by a KeyboardInterrupt event, at which point if the job ID
+        was retrieved from the server, the job will be accessible via engine.jobs.
+
+        Args:
+            program (strawberryfields.program.Program): A program instance to be executed remotely.
+
+        Returns:
+            (list): A list representing the result samples
+
+        Raises:
+            Exception: In case a job could not be submitted or completed.
+            TypeError: In case a job is already queued and a user is trying to submit a new job.
+        """
+        if self.jobs:
+            raise TypeError("A job is already queued. Please reset the engine and try again.")
+
+        blackbird_code = to_blackbird(program)
+        job_content = self.generate_job_content(blackbird_code=blackbird_code, **kwargs)
+        job = self._queue_job(job_content)
+
+        try:
+            while not job.is_failed and not job.is_complete:
+                job.reload()
+                sleep(self.API_DEFAULT_REFRESH_SECONDS)
+        except KeyboardInterrupt:
+            if job.id:
+                print("Job {} is queued in the background.".format(job.id.value))
+            else:
+                raise Exception("Job was not sent to server. Please try again.")
+
+        if job.is_failed:
+            # TODO: Add failure details here, and use a better exception.
+            raise Exception("Job execution failed. Please try again.")
+        elif job.is_complete:
+            job.result.manager.get()
+            return job.result.result.value
```

Comment:
Should this return the `job` instance, so that the user can manually check the status/reload the job/get the results when complete? See http://platform-docs/research/STRAW-92-engine-design.html#id1

Reply:
We should potentially rethink this in terms of how it applies to all engines. `_run_program` in `LocalEngine` and `BaseEngine` in fact, both return a list of applied commands, which is not applicable to this class. Currently, `StarshipEngine.jobs` has a list of jobs that are running, and can be checked or reloaded there. Async implementation, for the time being, is implemented as a `KeyboardInterrupt`, although I should add a flag for this (or possibly a different method, as the research doc suggests) so that a user can intentionally queue a job asynchronously without input.

Reply:
Tbh, I'm not entirely sure why `_run_program` needs to return a list of applied commands; it looks like `LocalEngine` doesn't even store the return value/do anything with it (unless I'm missing something).

Yep, it would be good to add a `run_async` in a later PR. Although, I quite like your `KeyboardInterrupt` idea, and even when we implement `run_async`, I still think we should keep both.

Reply:
`_run_program` could be removed from the base class and made a LocalEngine method only. `BaseEngine` is a very preliminary design, so if it makes sense to change the @abc.abstractmethods it defines, go for it.

Reply:
It may be that local and remote engines don't really have that much in common.

Reply:
@smite Yes true, originally I was trying to ensure a good fit between the two. They do have certain things in common obviously -- I guess this will get more interesting when/if we introduce even more engines and backends for each engine.